### ISSUE TITLE
Should changes in the state on getStateForAction unmount and mount screens?

### ISSUE THREAD

**Issue by otaviogaiao** [react-navigation/react-navigation#5197]
###Environment

react-navigation: 2.18.1
expo-sdk: 30
yarn: 1.9.4
node: 9.10.0

### Problem

I have an app with a main bottomTabNavigator, that contains 4 other navigators. Some of these navigators also have others navigators. Here is a diagram to try to explain better:

<img width="474" alt="screen shot 2018-11-01 at 14 28 12" src="https://user-images.githubusercontent.com/3965485/47868211-729c6680-dde2-11e8-9974-c7a5a2eaded7.png">

The user can perform login from both Stack D and Stack F, and when that happens I need to reset the state and unmount any screen in the other swithNavigator. 

I read the docs and tried to use the method getStateForAction inside my TabNavigator (since its the big parent).

Here is what I did:

```
//TabNavigator.js

export default navigator;

const defaultGetStateForAction = navigator.router.getStateForAction;

navigator.router.getStateForAction = (action, state) => {
  console.log(action, state);
  if (
    (state && action.type === LOGIN_PERFIL) ||
    action.type === LOGIN_INGRESSOS
  ) {
    let routes = [...state.routes];
    routes[2].index = 0;
    routes[3].index = 0;
    console.log(JSON.stringify(routes));
    return {
      ...state,
      routes
    };
  }
  return defaultGetStateForAction(action, state);
};
```

And on my LoginScreen I do:

```
              this.props.navigation.navigate("StackE"); //or StackC
              this.props.navigation.dispatch({
                type: LOGIN_PERFIL
              });
```

With this I reset only the current switch Navigator, but nothing happens with the other one. And Looking at the state that I get at getStateForAction, I see that the index for the switchNavigator is correctly changed.  Shouldn't the screen unmount when that happens?

I found nothing on the docs or the web related to that. Also no one could answer that on Discord.

Any help is appreciated.

**Comment by brentvatne** [comment]
hi! thanks for trying to provide a bunch of info - it's a bit hard to follow still, I'm happy to look at this if you provide a runnable example on https://snack.expo.io and steps to run through it

**Comment by otaviogaiao** [comment]
@brentvatne I created a small snack, I hope it helps identify an issue or at least points me what I am doing wrong.  https://snack.expo.io/@baladapp/navigation_error

Also, I recorded a video showing how to reproduce the issue.


![giphy](https://user-images.githubusercontent.com/3965485/48062648-c7a4f780-e1a9-11e8-9c88-f5af5587bfc4.gif)

**Comment by otaviogaiao** [comment]
@eriveltonelias this snack you showed and the docs are extremely simple. They don't fit what I need.
I need to be able to authenticate from two different tabs.

I tried to customize my router as showed in the first post (following the docs), but changing the state there doesn't unmount the screens. Thats my question

**Comment by eriveltonelias** [comment]
@otaviogaiao 

on the example that I sent, you can add more route on your TabNavigator and for each one add Stack Navigator and to open your login screen you can do the following: 

`this.props.navigation.navigate('Auth');`

**Comment by otaviogaiao** [comment]
@eriveltonelias the thing is: I can't have my Auth screen showing on the tab component. And looking at the docs, there doesn't seem to be a way to hide a specify route from the tab bar.

BUT the main issue persists, changes in the state inside **getStateForAction** shouldn't unmount and mount screens?? It is the code shown in the first post:

//TabNavigator.js

```
export default navigator;

const defaultGetStateForAction = navigator.router.getStateForAction;

navigator.router.getStateForAction = (action, state) => {
  console.log(action, state);
  if (
    (state && action.type === LOGIN_PERFIL) ||
    action.type === LOGIN_INGRESSOS
  ) {
    let routes = [...state.routes];
    routes[2].index = 0;
    routes[3].index = 0;
    console.log(JSON.stringify(routes));
    return {
      ...state,
      routes
    };
  }
  return defaultGetStateForAction(action, state);
};
```
